kips/kip-2.md

```diff
@@ -0,0 +1,462 @@
+  mapping(string => uint256[]) dataManyUint256;
+  mapping(string => bool[]) dataManyBool;
+}
+```
+
+**Proxy**
+
+Base class for all upgradeable contracts. Upgradeability is based on the [OpenZepellin Proxy + EternalStorage](https://github.com/OpenZeppelin/openzeppelin-labs/tree/master/upgradeability_using_eternal_storage) pattern.
+
+```solidity
+/**
+ * Upgradeable contracts base class
+ */
+contract Proxy is EternalStorage {
+  /**
+  * @dev This event will be emitted every time the implementation gets upgraded
+  * @param implementation representing the address of the upgraded implementation
+  */
+  event Upgraded(address indexed implementation);
+
+  /**
+   * Constructor.
+   */
+  constructor (address _implementation) public {
+    require(_implementation != address(0), 'implementation must be valid');
+    dataAddress["implementation"] = _implementation;
+  }
+
+  /**
+  * @dev Get the address of the implementation where every call will be delegated.
+  * @return address of the implementation to which it will be delegated
+  */
+  function getImplementation() public view returns (address) {
+    return dataAddress["implementation"];
+  }
+
+  /**
+   * @dev Point to a new implementation.
+   * This is internal so that descendants can control access to this in custom ways.
+   */
+  function setImplementation(address _implementation) internal {
+    require(_implementation != address(0), 'implementation must be valid');
+    require(_implementation != dataAddress["implementation"], 'already this implementation');
+
+    dataAddress["implementation"] = _implementation;
+
+    emit Upgraded(_implementation);
+  }
+
+  /**
+  * @dev Fallback function allowing to perform a delegatecall to the given implementation.
+  * This function will return whatever the implementation call returns
+  */
+  function () payable external {
+    address _impl = getImplementation();
+    require(_impl != address(0), 'implementation not set');
+
+    assembly {
+      let ptr := mload(0x40)
+      calldatacopy(ptr, 0, calldatasize)
+      let result := delegatecall(gas, _impl, ptr, calldatasize, 0, 0)
+      let size := returndatasize
+      returndatacopy(ptr, 0, size)
+      switch result
+      case 0 { revert(ptr, size) }
+      default { return(ptr, size) }
+    }
+  }
+}
+```
+
+**IACL, ACL, AccessControl**
+
+Contracts for handling access control. 
+
+At present the `ACL` stores the list of system admins and provides mechanisms for updating that list. The `AccessControl` class is how other contracts will make use of the ACL. Admins can add and remove other admins, but there will always be atleast 1 admin in the system.
+
+The ACL address is set once during construction but this could easily be changed to using an ENS lookup if we want to make the ACL itself replaceable in future.
+
+_Note: I just copied the ACL code from an existing project. There could be a simpler solution for admin management if this seems too complex._
+
```

Comment:
The ACL is relatively simple. My concern is the additional security risk. Am I correct to understand that if one of the admin's private key is compromises, the hacker can upgrade the whole contract and steal user's pot?

Reply:
Yes.

Reply:
But this could be mitigated by e.g. requiring 2+ admins to sign-off on an upgrade.